### PR TITLE
[GSOC 2025]: Github action for translation tracker

### PR DESCRIPTION
**Pull request by Divyansh013**
- test-local file detects local changes
- index.js does manual triggering for issue creation
- action works on all "push" events
- outdated banner addition on missing translation pages

### CODE REVIEW

**Comment by ksen0** [comment]
Hi @Divyansh013 , this is looking great!

Based on checking the current issue created by this (https://github.com/Divyansh013/p5.js-website-new/issues/11) I have a couple of minor suggestions:

* Move [🔍 Compare changes](https://github.com/Divyansh013/p5.js-website-new/compare/6ab310aaf0ffdc1be7f1afa4dd535fdff7b7325f...94159c8e1828541b0aeb6a803774f8dbd322d786) to the Quick Links section, because it's not file-specific.
* Potentially: next to this **one link** to "Compare changes", could add a note like: _ℹ️ You can see the changes in a specific file by going the the "Compare changes" link above, clicking on the "Files changed" tab, and navigating to the specific file._ 
* Move "Change Type: English file was updated. Spanish (Español), Hindi (हिन्दी), Korean (한국어), Chinese Simplified (简体中文) translations may be outdated." into the Timeline - right now this is very far separated, but seems related?

File: .github/actions/translation-tracker/index.js [hunk 1]
```diff
@@ -0,0 +1,646 @@
...
+    console.log('🧪 Using provided test files for local testing');
+    return testFiles.filter(file => 
+      file.startsWith('src/content/examples/en') && file.endsWith('.mdx')
+    );
+  }
+
+  try {
+    const gitCommand = process.env.GITHUB_EVENT_NAME === 'pull_request' 
+      ? 'git diff --name-only HEAD~1 HEAD'
+      : 'git diff --name-only HEAD~1 HEAD';
```
Comment (lirenjie95) [inline .github/actions/translation-tracker/index.js]: Two options are actually the same. Maybe typo here?

File: .github/actions/translation-tracker/index.js [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+          date: new Date(data[0].commit.committer.date),
+          message: data[0].commit.message,
+          author: data[0].commit.author.name,
+          url: data[0].html_url
+        };
+      }
+      
+      return null;
+    } catch (error) {
+      // Fallback to main branch if current branch fails
```
Comment (lirenjie95) [inline .github/actions/translation-tracker/index.js:113]: I'm wondering what will happen if the main branch also fails?

File: .github/actions/translation-tracker/index.js [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+    return false;
+  }
+}
+
+
+function getFileModTime(filePath) {
+  try {
+    return fs.statSync(filePath).mtime;
+  } catch (error) {
+    return null;
```
Comment (lirenjie95) [inline .github/actions/translation-tracker/index.js:515]: Add log here might be helpful for debugging.

File: .github/actions/translation-tracker/index.js [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+      console.error(`❌ Error creating multi-language issue:`, error.message);
+      return null;
+    }
+  }
+
+  /**
+   * Format the issue body with helpful information
+   */
+  formatIssueBody(englishFile, language, commitInfo) {
+    const translationPath = englishFile.replace('/en/', `/${language}/`);
```
Comment (lirenjie95) [inline .github/actions/translation-tracker/index.js]: If the directory structure changes or there are multiple occurrences of ·/en/· in the path (correct me if there are any situations that I didn't catch), this could lead to errors or unintended behavior.

File: .github/actions/translation-tracker/index.js (hunk 1 above)
Comment (Divyansh013) [reply .github/actions/translation-tracker/index.js]: indeed. What was originally intended here:  
```
'git diff --name-only origin/main...HEAD'  // Compare with base branch for PRs
      : 'git diff --name-only HEAD~1 HEAD'; // and with last commir for pushes
```
thoughts about this?

File: .github/actions/translation-tracker/index.js (hunk 2 above)
Comment (Divyansh013) [reply .github/actions/translation-tracker/index.js:113]: in that case we can log out the error ? or is there something I am missing out?

**Comment by Divyansh013** [comment]
@ksen0 requesting your review here

File: src/components/OutdatedTranslationBanner/index.astro [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    message: '이 페이지는 영어 버전과 비교하여 최신 상태가 아닙니다.',
+    viewEnglish: '영어 페이지 보기',
+    contribute: '번역에 기여하기',
+  },
+  'zh-Hans': {
+    title: '此翻译可能已过期',
+    message: '与英文版本相比，此页面不是最新的。',
+    viewEnglish: '查看英文页面',
+    contribute: '参与翻译',
+  },
```
Comment (lirenjie95) [inline src/components/OutdatedTranslationBanner/index.astro:42]: The Chinese translation here is pretty good. I will take a look of the remaining GitHub Actions related stuff when I have time.

File: .github/actions/translation-tracker/README.md [hunk 6]
```diff
@@ -0,0 +1,266 @@
+# p5.js Translation Tracker
+
+A comprehensive GitHub Action and command-line tool to track translation status across multiple languages in the p5.js website examples.
```
Comment (ksen0) [inline .github/actions/translation-tracker/README.md]: (Edited - see below)

File: .github/actions/translation-tracker/README.md [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+- **Week 1**: Basic file-based change detection using Git
+- **Week 2**: GitHub API integration for accurate commit tracking and automated issue creation
+- **Week 3**: Multi-language support with single issues per file and manual scanning capabilities
+
+## 📋 Features
+
+### ✅ Week 1 Features (File-based tracking)
+- Git-based change detection for English example files
+- File modification time comparison
+- Support for all languages: Spanish (es), Hindi (hi), Korean (ko), Chinese Simplified (zh-Hans)
```
Comment (ksen0) [inline .github/actions/translation-tracker/README.md]: Minor: a shorter README is easier to maintain, so please remove the duplicated info. (The supported languages are listed below already in a very readable format)

File: .github/actions/translation-tracker/README.md [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+### ✅ Week 2 Features (GitHub API integration)
+- Real commit-based comparison using GitHub API
+- Automated GitHub issue creation for outdated translations
+- Enhanced issue templates with helpful links and timelines
+- Branch detection (auto-detects current branch)
+- Backward compatibility with Week 1
+
+### ✅ Week 3 Features (Multi-language & refinement)
+- **Single issue per file**: Creates one issue covering all affected languages instead of separate issues per language
+- **Enhanced labeling**: Uses "needs translation" base label + specific language labels (e.g., "lang-es", "lang-ko")
```
Comment (ksen0) [inline .github/actions/translation-tracker/README.md]: (Edited - see above - mainly a very practical intro would be great in this README)

File: .github/actions/translation-tracker/README.md [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+- `GITHUB_TOKEN`: Required for Week 2+ features (API access and issue creation)
+- `GITHUB_REPOSITORY`: Auto-detected in GitHub Actions (format: `owner/repo`)
+- `GITHUB_EVENT_NAME`: Auto-detected in GitHub Actions
+- `GITHUB_REF_NAME`: Auto-detected branch name
+
+### Options Object
+
+```javascript
+{
+  enableWeek2: boolean,      // Force Week 2 mode
```
Comment (ksen0) [inline .github/actions/translation-tracker/README.md]: That seems probably not applicable anymore

File: src/components/OutdatedTranslationBanner/index.astro [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    contribute: 'अनुवाद में योगदान दें',
+  },
+  'es': {
+    title: 'Esta traducción podría estar desactualizada',
+    message: 'Esta página no está actualizada en comparación con la versión en inglés.',
+    viewEnglish: 'Ver versión en inglés',
+    contribute: 'Contribuir a la traducción',
+  },
+  'ko': {
+    title: '이 번역은 오래되었을 수 있습니다',
```
Comment (ksen0) [inline src/components/OutdatedTranslationBanner/index.astro:32]: Hi @hana-cho! When you have time, another set of Korean translations for your review. Thanks so much!

File: src/components/OutdatedTranslationBanner/index.astro [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+const copyByLocale: Record<string, { title: string; message: string; viewEnglish: string; contribute: string; }> = {
+  'hi': {
+    title: 'यह अनुवाद पुराना हो सकता है',
+    message: 'यह पृष्ठ अंग्रेज़ी संस्करण की तुलना में अद्यतन नहीं है।',
+    viewEnglish: 'अंग्रेज़ी संस्करण देखें',
+    contribute: 'अनुवाद में योगदान दें',
+  },
+  'es': {
+    title: 'Esta traducción podría estar desactualizada',
```
Comment (ksen0) [inline src/components/OutdatedTranslationBanner/index.astro:26]: Hi @marioguzzzman! If/when you have time, could you check these Spanish translations please? Thanks so much!

File: .github/actions/translation-tracker/README.md (hunk 6 above)
Comment (ksen0) [reply .github/actions/translation-tracker/README.md]: Actually, if this is for GSoC documentation, please feel free to keep the log, but maybe at the top of the file have a very short README of the tool? Just keeping separate the description of the action (for future maintenance) and the GSoC documentation would be great. For example, I think a good place to start this readme is to describe the overall flow, the format of the `translation-status` files, how they are produced/used.. The log you can of course keep as you have, it's just important to also have the summary for the tool over all 🙏

File: src/layouts/ExampleLayout.astro [hunk 12]
```diff
@@ -42,6 +45,49 @@ const relatedReferences =
...
 
 const { Content } = await example.render();
+
+// Determine if this example is marked outdated/missing for the current locale
+let showBanner = false;
+// Build English URL by manually replacing the locale prefix
+let englishUrl = Astro.url.pathname;
+if (currentLocale !== 'en') {
+  // Replace /{locale}/ with / to get English URL
+  englishUrl = englishUrl.replace(`/${currentLocale}/`, '/');
```
Comment (lirenjie95) [inline src/layouts/ExampleLayout.astro:55]: Question here: have you tried using `zh-Hans`? It is a different use case from `es`, `hi` and `ko`. I’m wondering if the replace operation will succeed.
PS: Please check all places with `currentLocale` and `replace()`

Comment (Divyansh013) [reply src/layouts/ExampleLayout.astro:55]: Yes have checked zh-Hans and it is working in expected way. I will check in all places

File: .github/actions/translation-tracker/README.md [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+## What It Does
+
+1. **Scans** English example files for changes
+2. **Compares** with translation files using Git commits
+3. **Creates** GitHub issues for outdated translations with:
+   - Diff snippets showing what changed
+   - Links to files and comparisons
+   - Action checklist for translators
+   - Proper labels (`needs translation`, `lang-es`, etc.)
+4. **Generates** manifest file for website banner system
```
Comment (ksen0) [inline .github/actions/translation-tracker/README.md:95]: Thanks so much for the README updates! As one last thing, please include here or ~line 112 a short explanation/example of the `translation-status` format.

File: .github/actions/translation-tracker/index.js [hunk 14]
```diff
@@ -0,0 +1,823 @@
...
+
+  /**
+   * Format the issue body with helpful information
+   */
+  formatIssueBody(englishFile, language, commitInfo) {
+    const translationPath = getTranslationPath(englishFile, language);
+    const englishCommit = commitInfo.english;
+    const translationCommit = commitInfo.translation;
+
+    return `## 🌍 Translation Update Needed
```
Comment (ksen0) [inline .github/actions/translation-tracker/index.js:302]: As a future improvement, this could be pulled out as a template file separate form the .js script.

File: .github/workflows/translation-sync.yml [hunk 15]
```diff
@@ -0,0 +1,38 @@
+name: Translation Sync Tracker
+
+on:
+  push:
+    branches: [main, week2]
```
Comment (ksen0) [inline .github/workflows/translation-sync.yml:5]: This is looking great and nearly ready for further testing. If we're now in step 1 of the below iterative deployment process, then to get to step 2, I think this should listen for push on branch **only** for now

1. Finalize this PR tested on your fork, document any further improvements, etc.
2. create a new PR that contains only the action. This PR would target the main branch, but the action would listen for translation changes on the staging branch. 
3. create a PR that targets the staging branch to add the UI. This will allow fine-tuning the tool.
4. Lastly, to make it go live, the process would be to marge the UI PR into main, and then change the action to listen on main rather than branch

File: .github/actions/translation-tracker/index.js [hunk 16]
```diff
@@ -0,0 +1,823 @@
...
+  } catch (error) {
+    console.error('❌ Error getting changed files:', error.message);
+    return [];
+  }
+}
+
+/**
+ * Scan all English example files (for manual scanning)
+ */
+function getAllEnglishExampleFiles() {
```
Comment (ksen0) [inline .github/actions/translation-tracker/index.js]: The biggest next step to improve here would be to generalize this code so it could work with not only "examples/" but also `tutorials/` and `text-detail/` (still not including reference) - I've tried to add comments where applicable!

File: .github/actions/translation-tracker/index.js [hunk 17]
```diff
@@ -0,0 +1,823 @@
...
+
+  // Get files to check
+  let filesToCheck;
+  if (testFiles) {
+    filesToCheck = getChangedFiles(testFiles);
+  } else if (isGitHubAction) {
+    filesToCheck = getChangedFiles();
+  } else {
+    console.log('📊 Scanning all English example files...');
+    filesToCheck = getAllEnglishExampleFiles();
```
Comment (ksen0) [inline .github/actions/translation-tracker/index.js]: Possible usage: `getAllExampleFiles(currentDirectory)` (where outside of the loop, `currentDirectory` is looping through, e .g., `["examples"]`). To keep the issue volume down during testing, I still think should should only focus on one of the 3 directories (examples), but it should be easily extensible to include others, without refactor. Locally, would alos be great to test on the other 2 directories to see what edge-cases may come up.

File: .github/actions/translation-tracker/index.js [hunk 18]
```diff
@@ -0,0 +1,823 @@
...
+  }
+  
+  if (filesToCheck.length === 0) {
+    if (isGitHubAction) {
+      console.log('✅ No English example files changed in this push');
```
Comment (ksen0) [inline .github/actions/translation-tracker/index.js:717]: "example" -> `currentDirectory` (or however you choose to name that var)

File: .github/actions/translation-tracker/index.js [hunk 19]
```diff
@@ -0,0 +1,823 @@
...
+    } else {
+      console.log('✅ No files to check');
+    }
+    return;
+  }
+  
+  console.log(`📝 Checking ${filesToCheck.length} English example file(s):`);
```
Comment (ksen0) [inline .github/actions/translation-tracker/index.js:724]: "example" -> `currentDirectory` (or however you choose to name that var)

File: .github/actions/translation-tracker/index.js [hunk 20]
```diff
@@ -0,0 +1,823 @@
...
+    filesToCheck, 
+    githubTracker, 
+    createIssues
+  );
+
+  // Detailed results
+  const { needsUpdate, missing, upToDate, issuesCreated } = translationStatus;
+  
+  console.log('\n📊 Translation Status Summary:');
+  console.log(`   🔄 Outdated: ${needsUpdate.length}`);
```
Comment (ksen0) [inline .github/actions/translation-tracker/index.js:738]: Adding a line here for something like  `console.log(`   🔍  Checked: ${translationDirectoriesList}`);`

File: .github/actions/translation-tracker/index.js [hunk 21]
```diff
@@ -0,0 +1,823 @@
...
+  } catch (writeErr) {
+    console.log(`\n⚠️  Could not write translation manifest: ${writeErr.message}`);
+  }
+}
+
+// Export for testing (simplified)
+module.exports = {
+  main,
+  getChangedFiles,
+  getAllEnglishExampleFiles,
```
Comment (ksen0) [inline .github/actions/translation-tracker/index.js:824]: As notes above, this would be great to generalize to not be just examples, but any similarly-structured dir on the website

File: .github/actions/translation-tracker/index.js [hunk 22]
```diff
@@ -0,0 +1,823 @@
...
+  }
+  
+  if (needsUpdate.length === 0 && missing.length === 0) {
+    console.log('\n✅ All translations are up to date!');
+  }
+
+  // Write manifest JSON for the site to consume
+  try {
+    const manifestDir = path.join(process.cwd(), 'public', 'translation-status');
+    const manifestPath = path.join(manifestDir, 'examples.json');
```
Comment (ksen0) [inline .github/actions/translation-tracker/index.js]: Can also use `currentDirectory` instead of hardcoding here

File: .github/actions/translation-tracker/index.js [hunk 23]
```diff
@@ -0,0 +1,823 @@
...
+    if (!fs.existsSync(manifestDir)) {
+      fs.mkdirSync(manifestDir, { recursive: true });
+    }
+    const examples = {};
```
Comment (ksen0) [inline .github/actions/translation-tracker/index.js]: `examples` potential rename: `directoryTranslationStatus` or something not specific to examples

File: .github/actions/translation-tracker/index.js [hunk 24]
```diff
@@ -0,0 +1,823 @@
...
+        upToDate: upToDateLangs,
+      };
+    }
+    const manifest = {
+      generatedAt: new Date().toISOString(),
+      branch: githubTracker ? githubTracker.currentBranch : null,
+      examples,
+    };
+    fs.writeFileSync(manifestPath, JSON.stringify(manifest, null, 2), 'utf8');
+    console.log(`\n🗂️  Wrote translation manifest: ${manifestPath}`);
```
Comment (ksen0) [inline .github/actions/translation-tracker/index.js]: Include current directory in output

File: .github/actions/translation-tracker/package.json [hunk 25]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "p5js-translation-tracker",
+  "version": "0.2.0",
+  "description": "GitHub Action to track translation status for p5.js examples and documentation",
```
Comment (ksen0) [inline .github/actions/translation-tracker/package.json]: "GitHub Action to track translation status for p5.js examples and documentation" -> "GitHub Action to track translation status for p5.js documentation"

File: .github/actions/translation-tracker/README.md [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+### GitHub Actions Workflow
+Create `.github/workflows/translation-sync.yml`:
+
+```yaml
+name: Translation Sync Tracker
+
+on:
+  push:
+    branches: [main, week2]
+    paths: ['src/content/examples/en/**']
```
Comment (ksen0) [inline .github/actions/translation-tracker/README.md:52]: This already looks generalized, which is great

File: .github/actions/translation-tracker/README.md [hunk 27]
```diff
@@ -0,0 +1,118 @@
+# p5.js Translation Tracker
+
+Automatically tracks translation status for p5.js website examples, creates GitHub issues for outdated translations, and shows banners on the website.
+
+## Features
+
+- Detects outdated/missing translations using Git commit comparison
```
Comment (ksen0) [inline .github/actions/translation-tracker/README.md]: Very minor: add a note here please that it currently focuses on Examples content

File: src/layouts/ExampleLayout.astro [hunk 28]
```diff
@@ -42,6 +45,49 @@ const relatedReferences =
...
+
+if (!englishUrl.startsWith('http')) {
+  englishUrl = `${Astro.url.origin}${englishUrl}`;
+}
+console.log('DEBUG: Current pathname:', Astro.url.pathname);
+console.log('DEBUG: English URL after manual replacement:', englishUrl);
+console.log('DEBUG: Current locale:', currentLocale);
+
+try {
+  const manifestPath = path.join(process.cwd(), "public", "translation-status", "examples.json");
```
Comment (ksen0) [inline src/layouts/ExampleLayout.astro:66]: I think at this stage, it makes sense to keep this here, but in the future (or now, optionally, if you prefer) also generalize it as a component usable in other layouts.

File: src/components/OutdatedTranslationBanner/index.astro (hunk 11 above)
Comment (marioguzzzman) [reply src/components/OutdatedTranslationBanner/index.astro:26]: The grammatically correct way would be to say "Ver la versión en inglés" but I think it works better and is more clean for an internet use context to leave it like: "Ver versión en inglés". So I would say the translation works correctly @ksen0!

Comment (ksen0) [reply src/components/OutdatedTranslationBanner/index.astro:26]: Thanks!
